keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,863 @@
+original cluster CIDR but allocates a /23 instead of a /24 to each node. They
+use the node selector to allocate these IPs only to the nodes with the higher
+capacity.
+
+#### Provision discontiguous ranges
+
+A user wants to create a cluster with 32 nodes each with a capacity to run 256
+pods. This means that each node needs a /24 pod CIDR range and they need a total
+range of /19. However, there aren't enough contiguous IPs in the user's network.
+They can find 4 free ranges of size /21 but no single contiguous /19 range.
+
+Using this enhancement, the user creates 4 different CIDR configurations each
+with a /21 range. The CIDR allocator allocates a /24 range from any of these /21
+ranges to the nodes and the user can now create the cluster.
+
+### Notes/Constraints/Caveats (Optional)
+
+A major precondition for this feature is to prevent churn on the IPs assigned to
+Nodes. Specifically, we want to prevent the controller from attempting to change
+a Node’s assigned PodCIDRs. For example, K8s will reject attempts to delete a
+CIDRRange that’s currently in use by a Node. This should prevent accidental
+broad-ranging changes to the cluster.
+
+### Risks and Mitigations
+
+-   Do racing controllers need a more sophisticated design? The current plan is
+    to plug in to the kube-controller-manager leader election.
+-   Can the controller install admission controls? This is a new pattern for
+    in-tree components.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### New Resource
+
+```
+type ClusterCIDRConfig struct {
+    # An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64")
+    # +required
+    CIDRBlock     string
+
+    # This defines which nodes the config is applicable to. An empty selector
+    # can be applied to any node.
+    # +optional
+    NodeSelector    *v1.LabelSelector
+
+    # Netmask size (e.g. 25 -> "/25") to allocate to a node.
+    # Users would have to ensure that the kubelet doesn't try to schedule
+    # more pods than are supported by the node's netmask (i.e. the kubelet's
+    # --max-pods flag)
+    # +required
+    PerNodeMaskSize int
+}
+```
+
+#### Expected Behavior
+
+-   Each node will be assigned up to one range from each `FamilyType`.
+    -   **TODO:** what is the tie-break if multiple ranges match
+-   An empty `NodeSelector` functions as a default that applies to all nodes.
+    This should be the fall-back and not take precedence if any other range
+    matches.
+-   `CIDRBlock`, and `PerNodeMaskSize` are immutable after creation.
+-   Reject any changes that would change existing node allocations:
```

Comment:
@aojea made an interesting point below -- we could adopt the model of PV/PVC where the controller puts finalizers on all the CIDRRangeConfigs that are in use. Attempting to delete a range in use would put it in a "terminated" phase so that no new Nodes can draw from that range, but the User must delete all nodes using that range before it is truly deleted.